samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/SecureDeliveryProctorio.java

```diff
@@ -0,0 +1,448 @@
+import org.sakaiproject.site.api.SiteService;
+import org.sakaiproject.tool.api.SessionManager;
+import org.sakaiproject.tool.assessment.data.dao.grading.SecureDeliveryData;
+import org.sakaiproject.tool.assessment.data.ifc.assessment.AssessmentMetaDataIfc;
+import org.sakaiproject.tool.assessment.data.ifc.assessment.PublishedAssessmentIfc;
+import org.sakaiproject.tool.assessment.data.ifc.assessment.SecureDeliveryModuleIfc;
+import org.sakaiproject.tool.assessment.facade.AgentFacade;
+import org.sakaiproject.tool.assessment.facade.PublishedAssessmentFacade;
+import org.sakaiproject.tool.assessment.services.PersistenceService;
+import org.sakaiproject.tool.assessment.shared.api.assessment.SecureDeliveryServiceAPI;
+import org.sakaiproject.tool.assessment.shared.api.assessment.SecureDeliveryServiceAPI.Phase;
+import org.sakaiproject.tool.assessment.shared.api.assessment.SecureDeliveryServiceAPI.PhaseStatus;
+import org.sakaiproject.user.api.User;
+import org.sakaiproject.user.api.UserDirectoryService;
+import org.sakaiproject.user.api.UserNotDefinedException;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.util.UriUtils;
+
+import javax.crypto.Mac;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+@Configuration
+@Slf4j
+public class SecureDeliveryProctorio implements SecureDeliveryModuleIfc {
+
+	// Encode as ASCII when communicating with Proctorio API
+	final private static String ENCODING = java.nio.charset.StandardCharsets.US_ASCII.toString();
+	// The property added to a site to enable and override options
+	final private static String SITE_PROPERTY = "proctorio";
+	// The header set by Proctorio
+	final private static String SESSION_PROPERTY = "x-proctorio";
+	// Default options if institution does not set
+	final private static String DEFAULT_OPTIONS = "recordvideo,linksonly";
+	// These are the options pulled from Proctorio API documentation. May need to be updated as their API changes over time.
+	final private static String[] VALID_OPTIONS = {
+			"recordvideo", "recordaudio", "recordscreen", "recordwebtraffic", "recordroomstart", 
+			"verifyvideo", "verifyaudio", "verifydesktop", "verifyidauto", "verifyidlive", "verifysignature",
+			"fullscreenlenient", "fullscreenmoderate", "fullscreensevere", "clipboard", "notabs", "linksonly",
+			"closetabs", "onescreen", "print", "downloads", "cache", "rightclick", "noreentry", "agentreentry", 
+			"calculatorbasic", "calculatorsci", "whiteboard"
+	};
+
+	private static String proctorioKey;
+	private static String proctorioSecret;
+	private static String proctorioUrl;
+	private static String proctorioEnabled;
+	private Cache<String, String> urlCache;
+
+	private UserDirectoryService userDirectoryService = ComponentManager.get(UserDirectoryService.class);
```

Comment:
new code should look at using SpringBeanAutowiringSupport

Reply:
i definitely tried using the AutoWired annotations